### PR TITLE
Spectrum refactor

### PR DESCRIPTION
**Pull request by andrewfullard**
### :pencil: Description

**Type:** :biohazard: `breaking change`

This PR moves as much of the spectrum out of the transport state as possible without also factoring out virtual packets. 

### :vertical_traffic_light: Testing

How did you test these changes?

- [x] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)

### :ballot_box_with_check: Checklist

- [ ] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

### CODE REVIEW

**Comment by andrewfullard** [comment]
Current issue: need to move and regenerate reference/regression data.

**Comment by andrewfullard** [comment]
Test failures are expected until https://github.com/tardis-sn/tardis-refdata/pull/85 and https://github.com/tardis-sn/tardis-regression-data/pull/8 are merged.

File: tardis/spectrum/base.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        "montecarlo_virtual_luminosity",
+        "spectrum",
+        "spectrum_virtual",
+        "spectrum_reabsorbed",
+        "spectrum_integrated",
+    ]
+
+    hdf_name = "spectrum"
+
+    def __init__(self, transport_state, spectrum_frequency):
```
Comment (andrewfullard) [inline tardis/spectrum/base.py:25]: Consider renaming `spectrum_frequency`

File: tardis/spectrum/base.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+from tardis.spectrum.formal_integral import IntegrationError
+from tardis.spectrum.spectrum import TARDISSpectrum
+from tardis.util.base import (
+    quantity_linspace,
+)
+
+
+class SpectrumSolver(HDFWriterMixin):
+    hdf_properties = [
+        "montecarlo_virtual_luminosity",
```
Comment (andrewfullard) [inline tardis/spectrum/base.py:16]: Should we be outputting this?

File: tardis/spectrum/base.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+            np.zeros_like(self.spectrum_frequency.value), "erg / s"
+        )  # should be init with v_packets_energy_hist
+        self._integrator = None
+        self.integrator_settings = None
+        self._spectrum_integrated = None
+
+    @property
+    def spectrum(self):
+        return TARDISSpectrum(
+            self.spectrum_frequency, self.montecarlo_emitted_luminosity
```
Comment (andrewfullard) [inline tardis/spectrum/base.py:38]: These luminosities are confusing

File: tardis/spectrum/base.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+from tardis.spectrum.spectrum import TARDISSpectrum
+from tardis.util.base import (
+    quantity_linspace,
+)
+
+
+class SpectrumSolver(HDFWriterMixin):
+    hdf_properties = [
+        "montecarlo_virtual_luminosity",
+        "spectrum",
```
Comment (andrewfullard) [inline tardis/spectrum/base.py:17]: `spectrum_real_packets`

File: tardis/spectrum/base.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            np.histogram(
+                self.transport_state.reabsorbed_packet_nu,
+                weights=self.transport_state.reabsorbed_packet_luminosity,
+                bins=self.spectrum_frequency,
+            )[0],
+            "erg / s",
+        )
+
+    @property
+    def montecarlo_emitted_luminosity(self):
```
Comment (andrewfullard) [inline tardis/spectrum/base.py:118]: luminosity density?

File: tardis/spectrum/base.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        luminosity_wavelength_filter = (
+            self.transport_state.reabsorbed_packet_nu > luminosity_nu_start
+        ) & (self.transport_state.reabsorbed_packet_nu < luminosity_nu_end)
+
+        return self.transport_state.reabsorbed_packet_luminosity[
+            luminosity_wavelength_filter
+        ].sum()
+
+    @classmethod
+    def from_config(cls, config, v_packets_energy_hist):
```
[the remark this answers is not in the document]
Comment (andrewfullard) [reply tardis/spectrum/base.py:182]: Or removal

File: tardis/simulation/base.py [hunk 7]
```diff
@@ -378,27 +383,39 @@ def iterate(self, no_of_packets, no_of_virtual_packets=0):
...
-        self.transport.run(
+        v_packets_energy_hist = self.transport.run(
             transport_state,
             time_explosion=self.simulation_state.time_explosion,
             iteration=self.iterations_executed,
             total_iterations=self.iterations,
             show_progress_bars=self.show_progress_bars,
         )
 
+        self.spectrum_solver = SpectrumSolver(
```
Comment (andrewfullard) [inline tardis/simulation/base.py:394]: Consider if this is the best place to init this class

File: tardis/transport/montecarlo/tests/test_base.py [hunk 8]
```diff
@@ -42,8 +42,8 @@ def test_hdf_transport(
     "output_energy",
     "nu_bar_estimator",
     "j_estimator",
-    "montecarlo_virtual_luminosity",
-    "packet_luminosity",
+    # "montecarlo_virtual_luminosity",
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/tests/test_base.py:45]: Just delete

File: tardis/spectrum/base.py (hunk 5 above)
Comment (andrewfullard) [reply tardis/spectrum/base.py:118]: it is not, it is the total luminosity per bin

File: tardis/simulation/base.py [hunk 9]
```diff
@@ -378,27 +383,39 @@ def iterate(self, no_of_packets, no_of_virtual_packets=0):
...
+            transport_state, self.transport.spectrum_frequency
+        )
+
+        self.spectrum_solver.integrator_settings = self.integrator_settings
+
+        self.spectrum_solver._montecarlo_virtual_luminosity.value[
```
[the remark this answers is not in the document]
Comment (andrewfullard) [reply tardis/simulation/base.py:400]: Again- this is luminosity total in each bin

File: tardis/spectrum/base.py (hunk 2 above)
Comment (Rodot-) [reply tardis/spectrum/base.py:16]: I don't see why not, might at least be useful for debugging

File: tardis/spectrum/base.py (hunk 1 above)
Comment (Rodot-) [reply tardis/spectrum/base.py:25]: perhaps frequency_grid?  I like names that imply a type if we're not going to annotate

**Review by Rodot- — DISMISSED** [review]
This looks good to me overall.

**Pull request merged**